saleor/order/calculations.py

```diff
@@ -0,0 +1,318 @@
+            )
+
+            subtotal += line_total.price_with_discounts
+
+        order.shipping_price = manager.calculate_order_shipping(order)
+        order.shipping_tax_rate = manager.get_order_shipping_tax_rate(
+            order, order.shipping_price
+        )
+        order.total = order.shipping_price + subtotal
+
+    except TaxError:
+        pass
+
+
+def _apply_tax_data(
+    order: Order, lines: Iterable[OrderLine], tax_data: TaxData
+) -> None:
+    """Apply all prices from tax data to order and order lines."""
+
+    def _quantize_price(net: Decimal, gross: Decimal) -> TaxedMoney:
+        currency = order.currency
+        return quantize_price(
+            TaxedMoney(net=Money(net, currency), gross=Money(gross, currency)),
+            currency,
+        )
+
+    order.total = _quantize_price(
+        net=tax_data.total_net_amount, gross=tax_data.total_gross_amount
+    )
+    order.shipping_price = _quantize_price(
+        net=tax_data.shipping_price_net_amount,
+        gross=tax_data.shipping_price_gross_amount,
+    )
+    order.shipping_tax_rate = tax_data.shipping_tax_rate
+
+    tax_lines = {line.id: line for line in tax_data.lines}
+    zipped_order_and_tax_lines = ((line, tax_lines[line.id]) for line in lines)
+
+    for (order_line, tax_line) in zipped_order_and_tax_lines:
+        order_line.unit_price = _quantize_price(
+            net=tax_line.unit_net_amount, gross=tax_line.unit_gross_amount
+        )
+        order_line.total_price = _quantize_price(
+            net=tax_line.total_net_amount, gross=tax_line.total_gross_amount
+        )
+        order_line.tax_rate = tax_line.tax_rate
+
+
+def _recalculate_order_discounts(order: Order, lines: Iterable[OrderLine]) -> None:
+    """Recalculate all order discounts and update order/lines prices."""
+    for line in lines:
+        line.undiscounted_unit_price = line.unit_price + line.unit_discount
+        line.undiscounted_total_price = (
+            line.undiscounted_unit_price * line.quantity
+            if line.unit_discount
+            else line.total_price
+        )
+
+    order_discounts = order.discounts.filter(type=OrderDiscountType.MANUAL)
+    for order_discount in order_discounts:
+        utils.update_order_discount_for_order(
+            order,
+            order_discount,
+        )
+
+
+def fetch_order_prices_if_expired(
+    order: Order,
+    manager: PluginsManager,
+    lines: Optional[Iterable[OrderLine]] = None,
+    force_update: bool = False,
+) -> Tuple[Order, Optional[Iterable[OrderLine]]]:
+    """Fetch order prices with taxes.
+
+    First calculate and apply all order prices with taxes separately,
+    then apply tax data as well if we receive one.
+
+    Prices can be updated only if force_update == True,
+    or if order price expiration time was exceeded
+    (which is settings.ORDER_PRICES_TTL if the prices are not invalidated).
+    """
+    with transaction.atomic():
```

Comment:
Since the entire logic is in the context manager, maybe it would be more readable to decorate the function with `transaction.atomic()`?

Reply:
We should push it down. Starting a transaction everytime this function is called makes a db connection which can add up to ~20-30 redundant queries per request

Reply:
I've done this in #9242 so let's leave it out of this PR